src/contexts/WalletContext/useWeb3Auth/index.ts

```diff
@@ -1,112 +1,147 @@
 
+        return { status: "disconnected" };
+      }
+
+      return { ...p, address: accounts[0] };
+    });
+  };
+
+  // //// PERSISTENT CONNECTION ////
   useEffect(() => {
-    const setupProviderInfo = async () => {
+    const adapter = web3Auth.cachedAdapter;
+    if (adapter) login(false);
+    //eslint-disable-next-line
+  }, []);
+
+  // //// USER INITIATED CONNECTION ////
+  const login = async (isNew = true) => {
+    try {
+      setState({ status: "loading" });
+
+      //there's always a need to initialize Web3Auth
+      await web3Auth.init();
+
+      const provider = isNew
+        ? await web3Auth.connectTo("torus-evm")
+        : web3Auth.provider;
+
       if (!provider) {
-        console.log("provider not initialized yet");
-        return;
+        // don't throw error on persistent connection
+        if (isNew) return;
+        throw new Error("Failed to connect to wallet");
+      }
+
+      const accounts = await provider.request<string[]>({
+        method: EIPMethods.eth_requestAccounts,
+      });
+      const hexChainId = await provider.request<string>({
+        method: EIPMethods.eth_chainId,
+      });
+
+      if (!(val(accounts) && val(hexChainId))) {
+        if (isNew) return;
+        throw new Error("Failed to connect to wallet");
       }
-      const rpc = new RPC(provider);
-      const chainId = await rpc.getChainId();
-      const address = await rpc.getAccounts();
-      setProviderInfo({
-        providerId: "web3auth-metamask",
-        logo: WEB3AUTH_LOGO,
-        chainId,
-        address,
+
+      provider.on("chainChanged", handleChainChange);
+      provider.on("accountsChanged", handleAccountsChange);
+
+      setState({
+        status: "connected",
+        address: accounts[0],
+        chainId: new Decimal(hexChainId).toString(),
+        provider: provider,
       });
-    };
-    if (provider) setupProviderInfo();
-  }, [provider, chainId]);
-
-  const login = async () => {
-    setIsLoading(true);
-    const web3Auth = await init();
-    if (!web3Auth) {
-      console.log("web3auth not initialized yet");
-      return;
+    } catch (err) {
+      setState({ status: "disconnected" });
+      //let caller of login() handle error
+      if (isNew) throw err;
+      //log for persistent connection
+      logger.error(err);
     }
-    const web3authProvider = await web3Auth.connect();
-    setProvider(web3authProvider);
-    setIsLoading(false);
   };
 
   const logout = async () => {
-    if (!web3auth) {
-      console.log("web3auth not initialized yet");
-      return;
-    }
     try {
-      await web3auth.logout();
-    } catch (e) {
-      console.log(e);
+      await web3Auth.logout({ cleanup: true });
+      web3Auth.provider?.off("chainChanged", handleChainChange);
+      web3Auth.provider?.off("accountsChanged", handleAccountsChange);
+      setState({ status: "disconnected" });
+    } catch (err) {
+      throw new Error("Failed to disconnect from wallet");
```

Comment:
```suggestion
      logger.error(err);
      throw new Error("Failed to disconnect from wallet");
```

Reply:
Throwing a generic error here is OK, but we still need to handle it in [WalletSubmission > line 38](https://github.com/AngelProtocolFinance/angelprotocol-web-app/blob/trim-web3auth/src/pages/Registration/Steps/WalletRegistration/WalletSubmission.tsx#L17).
The "handling" should probably be logging + displaying the error using the `ErrorContext.handleError`.